src/Web/Scotty/Internal/Types.hs

```diff
@@ -0,0 +1,240 @@
+    , routes :: [Middleware m]
+    , handler :: ErrorHandler e m
+    , routeOptions :: RouteOptions
+    }
+
+instance Default (ScottyState e m) where
+    def = ScottyState [] [] Nothing def
+
+addMiddleware :: Wai.Middleware -> ScottyState e m -> ScottyState e m
+addMiddleware m s@(ScottyState{middlewares = ms}) = s{middlewares = m : ms}
+
+addRoute :: Middleware m -> ScottyState e m -> ScottyState e m
+addRoute r s@(ScottyState{routes = rs}) = s{routes = r : rs}
+
+addHandler :: ErrorHandler e m -> ScottyState e m -> ScottyState e m
+addHandler h s = s{handler = h}
+
+updateMaxRequestBodySize :: RouteOptions -> ScottyState e m -> ScottyState e m
+updateMaxRequestBodySize RouteOptions{..} s@ScottyState{routeOptions = ro} =
+    let ro' = ro{maxRequestBodySize = maxRequestBodySize}
+    in  s{routeOptions = ro'}
+
+newtype ScottyT e m a = ScottyT {runS :: State (ScottyState e m) a}
+    deriving (Functor, Applicative, Monad)
+
+------------------ Scotty Errors --------------------
+data ActionError e
+    = Redirect Text
+    | Next
+    | Finish
+    | ActionError Status e
+
+-- | In order to use a custom exception type (aside from 'Text'), you must
+-- define an instance of 'ScottyError' for that type.
+class ScottyError e where
+    stringError :: String -> e
+    showError :: e -> Text
+
+instance ScottyError Text where
+    stringError = pack
+    showError = id
+
+instance ScottyError e => ScottyError (ActionError e) where
+    stringError = ActionError status500 . stringError
+    showError (Redirect url) = url
+    showError Next = pack "Next"
+    showError Finish = pack "Finish"
+    showError (ActionError _ e) = showError e
+
+type ErrorHandler e m = Maybe (e -> ActionT e m ())
+
+data ScottyException = RequestException BS.ByteString Status deriving (Show, Typeable)
+
+instance Exception ScottyException
+
+------------------ Scotty Actions -------------------
+type Param = (Text, Text)
+
+type File = (Text, FileInfo ByteString)
+
+data ActionEnv = Env
+    { getReq :: Request
+    , getParams :: [Param]
+    , getBody :: IO ByteString
+    , getBodyChunk :: IO BS.ByteString
+    , getFiles :: [File]
+    }
+
+data RequestBodyState
+    = BodyUntouched
+    | BodyCached ByteString [BS.ByteString] -- whole body, chunks left to stream
+    | BodyCorrupted
+
+data BodyPartiallyStreamed = BodyPartiallyStreamed deriving (Show, Typeable)
+
+instance E.Exception BodyPartiallyStreamed
+
+data Content
+    = ContentBuilder Builder
+    | ContentFile FilePath
+    | ContentStream StreamingBody
+
+data ScottyResponse = SR
+    { srStatus :: Status
+    , srHeaders :: ResponseHeaders
+    , srContent :: Content
+    }
+
+instance Default ScottyResponse where
+    def = SR status200 [] (ContentBuilder mempty)
+
+newtype ActionT e m a = ActionT {runAM :: ExceptT (ActionError e) (ReaderT ActionEnv (StateT ScottyResponse m)) a}
+    deriving newtype (Functor, Applicative, Monad, MonadIO)
+    deriving (Semigroup, Monoid) via (Ap (ActionT e m) a)
```

Comment:
Deriving `Semigroup`, `Monoid`, and `Monad`.

Reply:
doesn't this rule out old GHCs? or are we dropping them (which is also fine)

Reply:
> drop ghc pre 8.10 from ci

I suppose this is intentional